core/src/main/scala/kafka/Kafka.scala

```diff
@@ -17,28 +17,59 @@
-      .ofType(classOf[String])
-    // This is just to make the parameter show up in the help output, we are not actually using this due the
-    // fact that this class ignores the first parameter which is interpreted as positional and mandatory
-    // but would not be mandatory if --version is specified
-    // This is a bit of an ugly crutch till we get a chance to rework the entire command line parsing
-    optionParser.accepts("version", "Print version information and exit.")
+    def getPropsFromArgs(args: Array[String]): Properties = {
+        if (args.exists(_.contains("s3-url"))) {
+            val roleInfo = args.find(_.startsWith("process.roles="))
+            if (roleInfo.isEmpty) {
+                throw new IllegalArgumentException("'--override process.roles=broker|controller' is required")
+            }
+            if (!args.exists(_.startsWith("node.id"))) {
+                throw new IllegalArgumentException(s"'--override node.id= ' is required")
+            }
+            if (!args.exists(_.startsWith("controller.quorum.voters"))) {
+                throw new IllegalArgumentException(s"'--override controller.quorum.voters=''' is required")
+            }
+            if (!args.exists(_.startsWith("listeners"))) {
+                throw new IllegalArgumentException(s"'--override listeners=''' is required")
+            }
+
+            roleInfo match {
+                case Some("process.roles=broker") =>
+                    if (!args.exists(_.startsWith("advertised.listeners"))) {
+                        throw new IllegalArgumentException(s"'--override advertised.listeners=''' is required")
+                    }
+                    return S3ShellPropUtil.autoGenPropsByCmd(args, "broker")
+                case Some("process.roles=controller") =>
+                    return S3ShellPropUtil.autoGenPropsByCmd(args, "controller")
+                case _ =>
+                    if (!args.exists(_.startsWith("advertised.listeners"))) {
+                        throw new IllegalArgumentException(s"'--override advertised.listeners=''' is required")
+                    }
+                    return S3ShellPropUtil.autoGenPropsByCmd(args, "broker,controller")
+            }
+        }
+
+        val optionParser = new OptionParser(false)
+        val overrideOpt = optionParser.accepts("override", "Optional property that should override values set in server.properties file")
+            .withRequiredArg()
+            .ofType(classOf[String])
+        // This is just to make the parameter show up in the help output, we are not actually using this due the
+        // fact that this class ignores the first parameter which is interpreted as positional and mandatory
+        // but would not be mandatory if --version is specified
+        // This is a bit of an ugly crutch till we get a chance to rework the entire command line parsing
+        optionParser.accepts("version", "Print version information and exit.")
```

Comment:
- Wrong indent here
- When modifying the original Kafka code, it's better to add the comment "AutoMQ for Kafka inject start/end" before and after the code to facilitate maintenance. You can find examples elsewhere.